File: src/dex.ts
/**
 * @module dex
 * Provides on-chain DEX pool liquidity guardrails to prevent AI agents from
 * executing trades in illiquid honeypot pools and suffering massive slippage.
 */

import { createPublicClient, http, parseAbi, type PublicClient } from "viem";
import { mainnet } from "viem/chains";

const v3Abi = parseAbi(["function liquidity() view returns (uint128)"] as const);

const v2Abi = parseAbi([
  "function getReserves() view returns (uint112 _reserve0, uint112 _reserve1, uint32 _blockTimestampLast)",
] as const);

export type DexGuardrailResult = {
  decision: "ALLOW" | "BLOCK";
  reason: string;
  pool: string;
  allowExecute: boolean;
};

function blockDex(pool: string, reason: string): DexGuardrailResult {
  return {
    decision: "BLOCK",
    reason,
    pool,
    allowExecute: false,
  };
}

function allowDex(pool: string, reason: string): DexGuardrailResult {
  return {
    decision: "ALLOW",
    reason,
    pool,
    allowExecute: true,
  };
}

export type CheckPoolV3Input = {
  rpc: string;
  pool: string;
  minLiquidity: bigint;
  /** @internal inject mock client for tests */
  __client?: Pick<PublicClient, "readContract">;
};

/**
 * Checks a Uniswap V3 (or clone) pool to ensure its active in-range liquidity
 * meets a minimum safe threshold before an agent executes a swap.
 */
export async function checkPoolV3(input: CheckPoolV3Input): Promise<DexGuardrailResult> {
  const { rpc, pool, minLiquidity } = input;

  if (typeof minLiquidity !== "bigint" || minLiquidity < 0n) {
    return blockDex(pool, `invalid minLiquidity ${String(minLiquidity)} — BLOCK (fail closed)`);
  }
  if (!/^0x[a-fA-F0-9]{40}$/.test(pool)) {
    return blockDex(pool, `invalid pool address ${pool} — BLOCK`);
  }
  if (typeof rpc !== "string" || rpc.trim() === "") {
    return blockDex(pool, "missing rpc — BLOCK (fail closed)");
  }

  const client =
    input.__client ??
    createPublicClient({
      chain: mainnet, // chain doesn't matter for raw RPC calls without chain assertions
      transport: http(rpc),
    });

  let activeLiquidity: bigint;
  try {
    activeLiquidity = await client.readContract({
      address: pool as `0x${string}`,
      abi: v3Abi,
      functionName: "liquidity",
    });
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return blockDex(
      pool,
      `failed to read liquidity from V3 pool ${pool} — BLOCK (fail closed): ${msg}`,
    );
  }

  if (activeLiquidity < minLiquidity) {
    return blockDex(
      pool,
      `active liquidity ${activeLiquidity.toString()} < required ${minLiquidity.toString()} on V3 pool ${pool} — BLOCK`,
    );
  }

  return allowDex(
    pool,
    `V3 pool ${pool} liquidity ${activeLiquidity.toString()} meets minimum requirements`,
  );
}

export type CheckPoolV2Input = {
  rpc: string;
  pool: string;
  minReserve0: bigint;
  minReserve1: bigint;
  /** @internal inject mock client for tests */
  __client?: Pick<PublicClient, "readContract">;
};

/**
 * Checks a Uniswap V2 (or clone) pool to ensure both token reserves
 * meet a minimum safe threshold before an agent executes a swap.
 */
export async function checkPoolV2(input: CheckPoolV2Input): Promise<DexGuardrailResult> {
  const { rpc, pool, minReserve0, minReserve1 } = input;

  if (typeof minReserve0 !== "bigint" || minReserve0 < 0n) {
    return blockDex(pool, `invalid minReserve0 ${String(minReserve0)} — BLOCK (fail closed)`);
  }
  if (typeof minReserve1 !== "bigint" || minReserve1 < 0n) {
    return blockDex(pool, `invalid minReserve1 ${String(minReserve1)} — BLOCK (fail closed)`);
  }
  if (!/^0x[a-fA-F0-9]{40}$/.test(pool)) {
    return blockDex(pool, `invalid pool address ${pool} — BLOCK`);
  }
  if (typeof rpc !== "string" || rpc.trim() === "") {
    return blockDex(pool, "missing rpc — BLOCK (fail closed)");
  }

  const client =
    input.__client ??
    createPublicClient({
      chain: mainnet,
      transport: http(rpc),
    });

  let reserve0: bigint;
  let reserve1: bigint;
  try {
    const reserves = await client.readContract({
      address: pool as `0x${string}`,
      abi: v2Abi,
      functionName: "getReserves",
    });
    reserve0 = reserves[0];
    reserve1 = reserves[1];
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return blockDex(
      pool,
      `failed to read reserves from V2 pool ${pool} — BLOCK (fail closed): ${msg}`,
    );
  }

  if (reserve0 < minReserve0) {
    return blockDex(
      pool,
      `reserve0 ${reserve0.toString()} < required ${minReserve0.toString()} on V2 pool ${pool} — BLOCK`,
    );
  }

  if (reserve1 < minReserve1) {
    return blockDex(
      pool,
      `reserve1 ${reserve1.toString()} < required ${minReserve1.toString()} on V2 pool ${pool} — BLOCK`,
    );
  }

  return allowDex(pool, `V2 pool ${pool} reserves meet minimum requirements`);
}

File: src/dex.rs
use crate::abi::decode_word_u128;
use crate::addressbook::is_valid_eth_address;
use crate::rpc::EvmRpcClient;
use crate::types::GuardrailResult;
use serde_json::json;

pub const V2_GET_RESERVES_SELECTOR: &str = "0x0902f1ac";
pub const V3_LIQUIDITY_SELECTOR: &str = "0x1a686502";

pub async fn check_pool_v2(
    client: &dyn EvmRpcClient,
    pool: &str,
    min_reserve0: u128,
    min_reserve1: u128,
) -> GuardrailResult {
    if !is_valid_eth_address(pool) {
        return GuardrailResult::block(format!("invalid pool address {} — BLOCK", pool));
    }

    let hex_data = match client.call(pool, V2_GET_RESERVES_SELECTOR).await {
        Ok(h) => h,
        Err(e) => {
            return GuardrailResult::block(format!(
                "failed to read reserves from V2 pool {} — BLOCK (fail closed): {}",
                pool, e
            ));
        }
    };

    let reserve0 = match decode_word_u128(&hex_data, 0) {
        Ok(r) => r,
        Err(e) => {
            return GuardrailResult::block(format!(
                "failed to decode reserve0 from V2 pool {}: {}",
                pool, e
            ));
        }
    };

    let reserve1 = match decode_word_u128(&hex_data, 1) {
        Ok(r) => r,
        Err(e) => {
            return GuardrailResult::block(format!(
                "failed to decode reserve1 from V2 pool {}: {}",
                pool, e
            ));
        }
    };

    if reserve0 < min_reserve0 {
        return GuardrailResult::block(format!(
            "reserve0 {} < required {} on V2 pool {} — BLOCK",
            reserve0, min_reserve0, pool
        ))
        .with_metadata(json!({
            "reserve0": reserve0.to_string(),
            "reserve1": reserve1.to_string(),
            "pool": pool
        }));
    }

    if reserve1 < min_reserve1 {
        return GuardrailResult::block(format!(
            "reserve1 {} < required {} on V2 pool {} — BLOCK",
            reserve1, min_reserve1, pool
        ))
        .with_metadata(json!({
            "reserve0": reserve0.to_string(),
            "reserve1": reserve1.to_string(),
            "pool": pool
        }));
    }

    GuardrailResult::allow(format!(
        "V2 pool {} reserves meet minimum requirements",
        pool
    ))
    .with_metadata(json!({
        "reserve0": reserve0.to_string(),
        "reserve1": reserve1.to_string(),
        "pool": pool
    }))
}

pub async fn check_pool_v3(
    client: &dyn EvmRpcClient,
    pool: &str,
    min_liquidity: u128,
) -> GuardrailResult {
    if !is_valid_eth_address(pool) {
        return GuardrailResult::block(format!("invalid pool address {} — BLOCK", pool));
    }

    let hex_data = match client.call(pool, V3_LIQUIDITY_SELECTOR).await {
        Ok(h) => h,
        Err(e) => {
            return GuardrailResult::block(format!(
                "failed to read liquidity from V3 pool {} — BLOCK (fail closed): {}",
                pool, e
            ));
        }
    };

    let active_liquidity = match decode_word_u128(&hex_data, 0) {
        Ok(l) => l,
        Err(e) => {
            return GuardrailResult::block(format!(
                "failed to decode liquidity from V3 pool {}: {}",
                pool, e
            ));
        }
    };

    if active_liquidity < min_liquidity {
        return GuardrailResult::block(format!(
            "active liquidity {} < required {} on V3 pool {} — BLOCK",
            active_liquidity, min_liquidity, pool
        ))
        .with_metadata(json!({
            "liquidity": active_liquidity.to_string(),
            "pool": pool
        }));
    }

    GuardrailResult::allow(format!(
        "V3 pool {} liquidity {} meets minimum requirements",
        pool, active_liquidity
    ))
    .with_metadata(json!({
        "liquidity": active_liquidity.to_string(),
        "pool": pool
    }))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::mock::MockRpcClient;
    use std::sync::Arc;

    #[tokio::test]
    async fn test_pool_v2_sufficient() {
        let reserves_hex = format!("0x{:0>64x}{:0>64x}{:0>64x}", 1000u64, 2000u64, 1u64);
        let mock = MockRpcClient {
            call_handler: Some(Arc::new(move |_, _| Ok(reserves_hex.clone()))),
            ..Default::default()
        };

        let res = check_pool_v2(
            &mock,
            "0xB4e16d0168e52d35CaCD2c6185b44281Ec28C9Dc",
            500,
            1000,
        )
        .await;
        assert!(res.allow_execute);
    }

    #[tokio::test]
    async fn test_pool_v2_insufficient() {
        let reserves_hex = format!("0x{:0>64x}{:0>64x}{:0>64x}", 300u64, 2000u64, 1u64);
        let mock = MockRpcClient {
            call_handler: Some(Arc::new(move |_, _| Ok(reserves_hex.clone()))),
            ..Default::default()
        };

        let res = check_pool_v2(
            &mock,
            "0xB4e16d0168e52d35CaCD2c6185b44281Ec28C9Dc",
            500,
            1000,
        )
        .await;
        assert!(!res.allow_execute);
        assert!(res.reason.contains("reserve0 300 < required 500"));
    }

    #[tokio::test]
    async fn test_pool_v3_sufficient() {
        let liquidity_hex = format!("0x{:0>64x}", 50_000_000u64);
        let mock = MockRpcClient {
            call_handler: Some(Arc::new(move |_, _| Ok(liquidity_hex.clone()))),
            ..Default::default()
        };

        let res = check_pool_v3(
            &mock,
            "0x88e6A0c2dDD26FEEb64F039a2c41296FcB3f5640",
            10_000_000,
        )
        .await;
        assert!(res.allow_execute);
    }
}
